org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (c) 2012, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.xbase.formatting;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.xtext.formatting2.regionaccess.ITextRegionAccess;
+import org.eclipse.xtext.nodemodel.ICompositeNode;
+import org.eclipse.xtext.nodemodel.ILeafNode;
+import org.eclipse.xtext.nodemodel.INode;
+import org.eclipse.xtext.parsetree.reconstr.impl.NodeIterator;
+import org.eclipse.xtext.xbase.lib.Functions.Function1;
+import org.eclipse.xtext.xbase.lib.IntegerRange;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+import org.eclipse.xtext.xbase.lib.ListExtensions;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Strings;
+import com.google.inject.Inject;
+
+/**
+ * @deprecated use {@link ITextRegionAccess}
+ */
+@Deprecated
+public class HiddenLeafAccess {
+	@Inject
+	private NodeModelAccess nodeModelAccess;
+	private int offset;
+
+	public HiddenLeafs getHiddenLeafsBefore(INode node) {
+		ILeafNode start = nodeModelAccess.findNextLeaf(node, l -> !l.isHidden());
+		List<ILeafNode> nodes = this.findPreviousHiddenLeafs(start);
+		if (start != null) {
+			int offset = 0;
+			if (nodes.isEmpty()) {
+				offset = start.getOffset();
+			} else {
+				offset = IterableExtensions.head(nodes).getOffset();
+			}
+			return newHiddenLeafs(offset, nodes);
+		} else {
+			int offset = 0;
+			if (node != null) {
+				offset = node.getOffset();
+			}
+			return new HiddenLeafs(offset);
+		}
+	}
+
+	protected HiddenLeafs newHiddenLeafs(int offset, List<ILeafNode> nodes) {
+		HiddenLeafs result = new HiddenLeafs(offset);
+		boolean trailing = true;
+		for (ILeafNode node : nodes) {
+			boolean comment = false;
+			int newLines = 0;
+			if (!Strings.isNullOrEmpty(node.getText())) {
+				for (Integer i : new IntegerRange(0, node.getText().length() - 1)) {
```

Comment:
There are lot of `@Pure` in the file that declares `IntegerRange`

Reply:
I guess you'll handle this in another MR? Don't see any changes now

Reply:
? replaced it with a for loop ?!?

Reply:
That also works :+1: 

What I meant by my comment is that the file that declares IntegerRange has already undergone the xtend->java process, but we forgot to remove the `@Pure annotations` so we should also get rid of those

Reply:
no integerrange always was java. and its in xbase lib. thus keep the pure for xtend (only) users

Reply:
I see, makes sense